datameta/security/tfaz.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2021 Universität Tübingen, DKFZ and EMBL for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+import secrets
+
+from cryptography.fernet import Fernet
+import pyotp
+import qrcode
+import qrcode.image.svg
+from sqlalchemy import and_
+from pyramid import threadlocal
+from pyramid.settings import asbool
+
+from .tokenz import hash_token
+from ..models import User, Session, TfaToken
+
+import logging
+log = logging.getLogger(__name__)
+
+
+def is_2fa_enabled():
+    """Check if the system has 2fa enabled.
+
+    Returns:
+        - the 2fa activation status
+    """
+    return asbool(threadlocal.get_current_registry().settings['datameta.tfa.enabled'])
+
+
+def get_2fa_crypto_function(settings=None):
+    """Get the encrypt/decrypt function based on the system's master key.
+
+    Returns:
+        - a Fernet crypto function object
+
+    """
+    if settings is None:
+        settings = threadlocal.get_current_registry().settings
+    encrypt_key = str(settings['datameta.tfa.encrypt_key'])
+    return Fernet(encrypt_key.encode())
+
+
+def generate_2fa_secret(db, settings=None):
```

Comment:
From what I understand, it's not entirely clear to me why the db parameter is passed to the function.

Reply:
This is an artifact from previous iterations. In the current state, no reason to pass it.